### PR TITLE
feat(cheatcodes): Record Account and Storage Access Cheatcodes

### PR DESCRIPTION
**Pull request by Inphi**
## Motivation

This follows up the great work @refcell and @jameswenzel have done on #6087. It implements a record and storage access cheatcode interface based on #6125.

## Solution

#6125 goes over the details of the implementation. However, the interface proposed has been tweaked to the following:
```
enum AccountAccessKind {
  Call, CallReturn, CallCode, StatiCall,
  Create,
  SelfDestruct,
  Resume,
}

struct ChainInfo { uint256 forkId; uint256 chainId; }
 
struct AccountAccess {
    ChainInfo chainInfo;
    AccountAccessKind kind;
    address account;
    address accessor;
    bool initialized;
    uint256 oldBalance;
    uint256 newBalance;
    bytes deployedCode;
    uint256 value;
    bytes data;
    bool reverted;
    StorageAccess[] storageAccesses;
}

struct StorageAccess {
    address account;
    bytes32 slot;
    bool isWrite;
    bytes32 previousValue;
    bytes32 newValue;
    bool reverted;
}

function startStateDiffRecording() external;
function stopAndReturnStateDiff() external returns (AccountAccess[] memory accesses);
```

Once `cheats.startStateDiffRecording()` is called, all state and account accesses will be recording _starting from the next context_, rather than the current one. For example, given the following code snippet:
```solidity
function run() {
  B b = new B();
  cheats.startStateDiffRecording();
  A a = new A(b);
  a.foo();
}

contract A {
  B _b;
  constructor(B b) { b = b }
  function foo() external { _b.foo(); }
}
```
The recorded accesses will contain the following:
- CREATE opcode on A
- CALL on a.foo
- CALL on b.foo

Which excludes the CREATE context for B's ctor because no AccountAccess record exists for the current context at the time recording was enabled. Without keeping track of all `AccountAccess` records even when recording is disabled, we cannot emit a well-defined `AccountAccess` for B's ctor.

### Control-Flow Linking
A context being recorded may be temporarily switched via sub-calls. Once control flow is returned to that context, it's important to maintain an ordering of subsequent accesses. This is accomplished through the `Resume` access kind. A `Resume` AccountAccess record contains storage accesses that occur between context switches. For example, given the following:
```solidity
function run() {
  cheats.startStateDiffRecording();
  a.foo();  
}

contract A {
  function foo() external {
    assembly { sstore(0, 0) }
    this.bar(); //
    assembly { sstore(1, 0) }
  }
  function bar() external {}
} 
```
We record the following:
- AccountAccess for the A.foo `Call`
  - StorageAccess for slot 0 store
- AccountAccess for the A.bar `Call`
- AccountAccess for the A.foo `Return`
  - StorageAccess for slot 1 store

Here, the `Return` allows users determine the context where the storage access occurred. The `Return` account access is only created if there were subsequent storage accesses made after a context switch. This interface only aims to provide enough information to track state changes, rather than offering a comprehensive control flow trace.

### CODE REVIEW

File: crates/cheatcodes/defs/src/vm.rs
Comment (mds1) [inline crates/cheatcodes/defs/src/vm.rs:148]: Should this struct and the `StorageAccess` struct include a `forkId` and/or `chainId`, for the rationale given [here](https://github.com/foundry-rs/foundry/issues/6125#issuecomment-1799049311)? Existing recording cheats don't include this, so I'm ok not including it, but in that case we probably should create an issue track a breaking change to all these cheats that adds this (cc @Evalir)

File: crates/cheatcodes/defs/src/vm.rs
Comment (mds1) [inline crates/cheatcodes/defs/src/vm.rs:157]: What does `initialized` mean exactly, e.g. what state affects this bool (nonce, code, storage, balance). And this is referring to whether or not it was initialized _before_ the access?

File: crates/cheatcodes/defs/src/vm.rs
Comment (mds1) [inline crates/cheatcodes/defs/src/vm.rs:161]: > This represents the balance change ignoring any reverts that may have occurred.

So this is really a *potential* new balance that is only accurate if the call did not revert, and not accurate if the call reverted?

File: crates/cheatcodes/defs/src/vm.rs
Comment (mds1) [inline crates/cheatcodes/defs/src/vm.rs:177]: So if address A delegatecalls to B, we'd have:
- `AccountAccess.kind = CALL`
- `AccountAccess.account = B`
- `StorageAccess.account = A`
- The slots all refer to storage in A

Is that correct? Should we make the `AccountAccessKind` more granular so it includes staticcall, delegatecall, and callcode, that way it's explicit instead of implicit what kind of call was made?

File: crates/cheatcodes/defs/src/vm.rs
Comment (mds1) [inline crates/cheatcodes/defs/src/vm.rs:45]:
```suggestion
        /// The account was created.
        Create,
        /// The account was selfdestructed.
```

**Review by mds1 — COMMENTED** [review]
I haven't reviewed the implementation, but have some questions around UX and behavior. Overall looks great though agree with the direction!

File: crates/cheatcodes/defs/src/vm.rs
Comment (Inphi) [reply crates/cheatcodes/defs/src/vm.rs:157]: This basically indicates whether the account existed just before it was accessed. The account exists if it either has code, a non-zero nonce or a non-zero balance.

File: crates/abi/abi/HEVM.sol
Comment (Inphi) [reply crates/abi/abi/HEVM.sol:90]: Depends on where the record was initiated. If you start recording just prior to `A.foo()`, then there wouldn't be any `AccountAccess.A` since A wasn't accessed. The first AccountAccess would be for B, with A as its _accessor_. So the returned array would be as you say, without the `AccountAccess.A`.

Yeah, I'm starting to agree with you on a new AccessKind needed for this return control flow. We do need to maintain the "chronological" order so `StorageAccess C, slot 3` relates with `StorageAccess B, slot 5`. But it won't be clear to users that the latter storage access isn't from a call.

OTOH, this kinda defeats the purpose of having a `storageAccesses` field for `AccountAccess` if we need a new kind to handle this case. I'm not exactly sure the right approach should be. But I lean towards completely decoupling storage accesses from account accesses, while maintaining ordering. So you have a the following state diff array:
- AccountAccess.B
- StorageAccess.B, slot 0
- AccountAccess.C
- StorageAccess.C, slot 3
- StorageAccess.B, slot 5

The main issue with this approach is that the user cannot tell the context `StorageAccess.B, slot 5` is for. This _should_ be evident from the `StorageAccess.account` field, but perhaps in cases like delegatecall it's a bit blurry.
This API depends on whether users need to know exactly which context an account access occurred.

File: crates/cheatcodes/defs/src/vm.rs
Comment (Inphi) [reply crates/cheatcodes/defs/src/vm.rs:161]: Correct. So if account A calls `B.foo{value: 1 ether}()` and then A later reverts, its `newBalance` will be 1 eth short.

File: crates/cheatcodes/defs/src/vm.rs
Comment (Inphi) [reply crates/cheatcodes/defs/src/vm.rs:169]: It's both. If either the current or parent context reverts, then `reverted` is set.
I'll clarify this in the documentation.

File: crates/abi/abi/HEVM.sol
Comment (mds1) [reply crates/abi/abi/HEVM.sol:90]: > Depends on where the record was initiated. If you start recording just prior to `A.foo()`, then there wouldn't be any `AccountAccess.A` since A wasn't accessed. The first AccountAccess would be for B, with A as its _accessor_. So the returned array would be as you say, without the `AccountAccess.A`.

Interesting, I'd expect this to have resulted in `AccountAccess.A` entry:
```solidity
contract MyTest is Test {
  function testAccesses() external {
    vm.recordStateDiff();
    a.foo(); // `foo` now calls `b`
    AccountAccess[] memory accesses = vm.getStateDiff();
  }
}
```
And I think it should, because users will want the storage accesses that took place in the `a.foo()` call also

---

Hmm, yea this is tricky. A "return control flow" seems ok but feels hacky. And I don't think Solidity is flexible enough to allow the cheat to return a flat array of `(AccountAccess | StorageAccess)[]` so the UX there may be clunky because I think you'd need it return a struct of:
```solidity
struct AccountOrStorageAccess {
  AccountAccess accountAccess; // empty if it was a storage access
  StorageAccess storageAccess; // empty if it was an account access
  bool isAccountAccess;
  bool isStorageAccess;
}
```

File: crates/cheatcodes/defs/src/vm.rs
Comment (refcell) [reply crates/cheatcodes/defs/src/vm.rs:148]: Just saw this comment, yes definitely agree with the inclusion of both the `forkId` and `chainId`. Even if the happy-path of using them is rarely hit, it's inexpensive to add this data to the object.

File: crates/cheatcodes/defs/src/vm.rs
Comment (refcell) [reply crates/cheatcodes/defs/src/vm.rs]: Fixed :)

File: crates/abi/abi/HEVM.sol
Comment (Inphi) [reply crates/abi/abi/HEVM.sol:90]: Let me clarify how recording occurs. In your code snippet, the `a.foo()` CALL will be recorded. What I was trying to explain is that state accesses in the same context a record begins are not recorded. That is the `sstore` in the following example won't be recorded:
```solidity
contract MyTest is Test {
  function testAccesses() external {
    vm.recordStateDiff();
    assembly { sstore(0x0, 0x0) }
    a.foo(); // `foo` now calls `b`
    AccountAccess[] memory accesses = vm.getStateDiff();
  }
}
```
The `a.foo()` call generates an `AccountAccess` record for only the `b` contract, but not `a` since that's the only account being accessed there. I guess, if there is a value transfer then it should generate two records for the credit and debit changes. Though this will have to be designated with a new type of account access or even a new Access type struct.

File: crates/cheatcodes/defs/src/vm.rs
Comment (refcell) [reply crates/cheatcodes/defs/src/vm.rs]: Fixed :)

File: crates/cheatcodes/defs/src/vm.rs
Comment (refcell) [reply crates/cheatcodes/defs/src/vm.rs]: Fixed this :)

File: crates/cheatcodes/defs/src/vm.rs
Comment (refcell) [reply crates/cheatcodes/defs/src/vm.rs]: Clarified this such that it uses "potential" and reads cleaner.

File: crates/cheatcodes/defs/src/vm.rs
Comment (refcell) [reply crates/cheatcodes/defs/src/vm.rs]: Clarified.

File: crates/abi/abi/HEVM.sol
Comment (Inphi) [reply crates/abi/abi/HEVM.sol:90]: In your example, is the `a.foo()` an actual CALL, i.e. calling an external A.foo function for example? If so, then I think we're on the same page. If it's just a jump, then that kinda breaks the CALL kind concept we have going.

Comment (mds1) [reply crates/abi/abi/HEVM.sol:90]: Oh sorry it was a CALL, I see how that is unclear

Comment (mds1) [inline crates/abi/abi/HEVM.sol:90]: We should rename this to `getRecordedStateDiff` to match `getLogs` and `getRecordedLogs`

File: crates/cheatcodes/spec/src/vm.rs
Comment (refcell) [inline crates/cheatcodes/spec/src/vm.rs:53]: nit, duplication of `sub-context` here?
```suggestion
        /// Synthetic access indicating the current context has resumed after a previous sub-context (AccountAccess).
```

File: crates/abi/abi/HEVM.sol
Comment (mds1) [inline crates/abi/abi/HEVM.sol:9]: These might be out of date? For example `forkId` and `chainId` are missing here

File: crates/cheatcodes/spec/src/vm.rs
Comment (mds1) [inline crates/cheatcodes/spec/src/vm.rs:49]: Just a typo on the capital H
```suggestion
        /// The account was called via staticcall.
        StaticCall,
        /// The account was created.
```

File: crates/abi/abi/HEVM.sol
Comment (mds1) [reply crates/abi/abi/HEVM.sol]: Just noting for completeness here that we discussed offline and agreed on adding a synthetic `Resume` access type to handle the "control flow returned to caller" access kind

File: crates/abi/abi/HEVM.sol
Comment (Inphi) [reply crates/abi/abi/HEVM.sol:9]: good catch.

**Review by mattsse — CHANGES_REQUESTED** [review]
some style nits,
defer to @mds1 and @DaniPopes

File: crates/abi/abi/HEVM.sol
Comment (mds1) [inline crates/abi/abi/HEVM.sol:8]: Can we include the enum definition in this file and that instead of `uint256 kind`?

File: crates/cheatcodes/spec/src/vm.rs
Comment (mds1) [inline crates/cheatcodes/spec/src/vm.rs:40]: Nit, but I think we should lowercase everything here and in the solidity definitions? e.g. `Delegatecall`, Callcode`, `Staticcall`, `Selfdestruct`? I feel like I don't usually see those words cased in the current way, but happy to go with whatever is consistent with the rest of the codebase/revm

Comment (Inphi) [reply crates/cheatcodes/spec/src/vm.rs:40]: leaving it capital cased for consistency. The `CallerMode` enums use capital case.

Comment (mds1) [reply crates/cheatcodes/spec/src/vm.rs:40]: Ah you're right this matches revm: https://github.com/bluealloy/revm/blob/aa5667764890bc554ea6392ea1e84e3e1991be6d/crates/interpreter/src/inner_models.rs#L63-L72

And Safe: https://github.com/safe-global/safe-contracts/blob/915045fcf7e62cc4564c1755b2676b8a2ab22b10/contracts/common/Enum.sol#L11

So, marking this resolved

File: crates/abi/abi/HEVM.sol
Comment (mds1) [reply crates/abi/abi/HEVM.sol:8]: Ah interesting, should be ok if we can't use the enum here since we can still use in `Vm.sol` in forge-std, so will defer to you + @mattsse / @Evalir here about the abigen issue

File: crates/cheatcodes/src/inspector.rs
Comment (Inphi) [reply crates/cheatcodes/src/inspector.rs]: I've updated the doc. Reproducing the revised comments:
>   /// Recorded account accesses (calls, creates) organized by relative call depth, where the
    /// topmost vector corresponds to accesses at the depth at which account access recording
    /// began. Each vector in the matrix represents a list of accesses at a specific call
    /// depth. Once that call context has ended, the last vector is removed from the matrix and
    /// merged into the previous vector.

File: crates/cheatcodes/src/inspector.rs
Comment (Inphi) [reply crates/cheatcodes/src/inspector.rs]: It should be vec![] (and clippy should detect this but doesn't). Fixed.

File: crates/cheatcodes/src/inspector.rs
Comment (Inphi) [reply crates/cheatcodes/src/inspector.rs]: hmm why's that? clippy doesn't flag either as problematic.

Comment (Inphi) [reply crates/cheatcodes/src/inspector.rs]: gotcha. Thanks for the context. The rest of the code uses the new style, so I'll revert it back to what it was originally.

File: crates/abi/abi/HEVM.sol
Comment (DaniPopes) [reply crates/abi/abi/HEVM.sol]: don't worry about it, crates/abi is deprecated and will be removed soon. It's currently only used for trace/log decoding

File: crates/cheatcodes/spec/src/vm.rs
Comment (DaniPopes) [inline crates/cheatcodes/spec/src/vm.rs:254]: nit: from UX perspective, would it be useful to have a "return accesses but don't stop", or alternatively "stop but don't clear existing accesses"? Does the current behaviour mimic the other record cheats?

Comment (Inphi) [reply crates/cheatcodes/spec/src/vm.rs:254]: This behavior does mimic other cheats like `getRecordedLogs`, which also stops log recording. I decided to rename it a bit differently so that it's clear that it also stops recording.
Other record cheats such as `record`/ `accesses` and `recordLogs` / `getRecordedLogs` don't have functions that  only stop without returning anything either.
I suppose users can always call `stopAndReturnStateDIff` without using the returned state diffs if they just want to stop recording.

File: crates/cheatcodes/src/inspector.rs
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:440]: This can be collapsed into `let (old_balance, initialized) = ... { (account, ...) } else { (...) }`

File: crates/cheatcodes/src/inspector.rs
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:449]: This doesn't work as you'd expect. U160::from panics on overflow, you want `Address::from_word(B256::from(target))`. Also this can be moved at the start of the block to remove duplication in `load_account`.

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/cheatcodes/src/inspector.rs
Comment (mattsse) [inline crates/cheatcodes/src/inspector.rs:155]: pedantic naming nit:

`recorded_account_diffs_stack`

File: crates/cheatcodes/src/inspector.rs
Comment (mattsse) [inline crates/cheatcodes/src/inspector.rs:454]:
```suggestion
                if let Some(last) = account_accesses.last_mut() {
```

**Review by mattsse — CHANGES_REQUESTED** [review]
I'd like to rename the variable

otherwise lgtm, pending @DaniPopes 

I still prefer `ref mut` because it makes it instantly clear to me when looking at the `Some` that this is a &mut without checking the rhs
but I don't mind &mut

File: crates/cheatcodes/src/inspector.rs
Comment (Inphi) [reply crates/cheatcodes/src/inspector.rs]: fixed.

**Review by mattsse — APPROVED** [review]

**Review by DaniPopes — APPROVED** [review]
lgtm

**Comment by mattsse** [comment]
gg
sending it

**Comment by refcell** [comment]
Wooot thank you @mattsse @DaniPopes 👑 👑 👑 👑

**Comment by mds1** [comment]
@Inphi Can you PR the cheats into forge-std's `VmSafe` in `Vm.Sol`, and the book also?

**Comment by Inphi** [comment]
> @Inphi Can you PR the cheats into forge-std's `VmSafe` in `Vm.Sol`, and the book also?

yup. Will do

**Pull request merged**
